### ISSUE TITLE
Kurento build with current gstreamer 1.0 (1.18) - webrtcendpoint 

### ISSUE THREAD

**Issue by ppisa** [Kurento/bugtracker#589]
## Prerequisites

These are MANDATORY, otherwise the issue will be automatically closed.

* [x] I agree to fill this issue template.
* [x] I have read the [Troubleshooting Guide] and [Support Instructions].

[Troubleshooting Guide]: https://doc-kurento.readthedocs.io/en/latest/user/troubleshooting.html
[Support Instructions]: https://github.com/Kurento/.github/blob/master/SUPPORT.md

## Issue description

webrtcendpoint is not found in gstreamer 1.0 1.18 registry

```[onmessage] Received message: {"id":"ERROR","message":"[Kurento] Exception: Cannot create gstreamer element: webrtcendpoint (Code:40108, Type:null, Data: {\"type\":\"MEDIA_OBJECT_NOT_AVAILABLE\"})"}
Kurento error: [Kurento] Exception: Cannot create gstreamer element: webrtcendpoint (Code:40108, Type:null, Data: {"type":"MEDIA_OBJECT_NOT_AVAILABLE"})
Assume that the other side stops after an error...
[stop]
[sendMessage] message: {"id":"STOP"}```

Cause is mismatch between `libwebrtcendpoint.so` library name and entry points `gst_plugin_kmswebrtcendpoint_get_desc`, `gst_plugin_kmswebrtcendpoint_register`.

Next links solve the problem

```libkmscore.so -> libkmscoreplugins.so
libkmselementsplugins.so
libkmselements.so -> libkmselementsplugins.so
libkmsfacedetector.so -> libfacedetector.so
libkmsfaceoverlay.so -> libfaceoverlay.so
libkmsimageoverlay.so -> libimageoverlay.so
libkmslogooverlay.so -> liblogooverlay.so
libkmsmovementdetector.so -> libmovementdetector.so
libkmsopencvfilter.so -> libopencvfilter.so
libkmsrecorderendpoint.so -> librecorderendpoint.so
libkmsrtpendpoint.so -> librtpendpoint.so
libkmswebrtcendpoint.so -> libwebrtcendpoint.so```

I have not working demo yet but there is some progress. I think that patch to align library names with entry point functions names is acceptable even for actual builds of Kurento based on dated gstreamer. Actual progress is till the point 

```
[onmessage] Received message: {"id":"PROCESS_SDP_ANSWER","sdpAnswer":"v=0\r\no=- 3826518882 3826518882 IN IP4 0.0.0.0\r\ns=Kurento Media Server\r\nc=IN IP4 0.0.0.0\r\nt=0 0\r\na=sendrecv\r\na=fingerprint:sha-256 DF:C3:48:00:A5:75:40:6A:F7:8D:69:56:0C:3F:F9:CB:EE:03:88:97:DF:68:35:95:1E:65:CC:D8:8D:3B:87:C4\r\na=ice-options:trickle\r\na=msid-semantic:WMS *\r\na=group:BUNDLE\r\nm=audio 0 UDP/TLS/RTP/SAVPF\r\na=sendrecv\r\na=mid:0\r\na=rtcp:9 IN IP4 0.0.0.0\r\na=rtcp-mux\r\na=setup:active\r\nm=video 0 UDP/TLS/RTP/SAVPF\r\na=extmap:4 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time\r\na=sendrecv\r\na=mid:1\r\na=rtcp:9 IN IP4 0.0.0.0\r\na=rtcp-mux\r\na=setup:active\r\n"}
[handleProcessSdpAnswer] SDP Answer from Kurento, process in WebRTC Peer
[handleProcessSdpAnswer] SDP Answer ready; start remote video
```

and `SDP Answer from KMS to browser` seems to be promising. 

On Kurento side, the progress is stuck on

```
(kurento-media-server:2284): GStreamer-CRITICAL **: 11:54:42.159: gst_bin_get_by_name: assertion 'GST_IS_BIN (bin)' failed
0:00:43.215763944  2284 0x7f264400a330 WARN     webrtctransportsink kmswebrtctransportsink.c:63:kms_webrtc_transport_sink_connect_elements: Cannot get srtpenc with name srtp-encoder
0:00:43.215775760  2284 0x7f264400a330 ERROR     kmswebrtctransport kmswebrtctransport.c:94:kms_webrtc_transport_new: SRTP plugin not available: dtlssrtpenc, dtlssrtpdec
``` 

So that is the next step to investigate.

## Context

Kurento requirement/dependency on gstreamer-1.5 libraries based on dated 1.8.x (my finding origin in 2016) version is long term blocker for BigBlueButton and OpenVidu and all related technologies combination with sane GNU/Linux releases which obtain demanded security updates and are based on more performant Linux kernels and libraries.

We use BigBlueButton on Ubuntu 16.4 and it allowed us to continue teaching even in forced distance form with strict relying on binary blobs from big vendors. But Ubuntu 16.04 is big obstacle for virtualization etc. I want to provide access to our development kits remotely to students and I started investigate how to use WebRTC for this and because permanent BigBluButton room seems as to big overhead for our loaded systems, I have decided to try underlying Kurento technology on my desktop system to share video access to the hardware with my students.

I have spent week with attempt to run Ubuntu 16.04 in LXC on Debian stable host, but without success. I have attempted to pull gstreamer 1.5 into my system, nightmare. So I decided to try gstreamer from upcoming Debian 11 on Debian 10 and test if there is chance to run Kurento based on gstreamer 1.0 1.80 (my build) and other regular Debian backports.

The results are partially promising. I replaced gstreamer 1.5 references in  kms-omni-build to 1.0 1.18 and Kurento build would not break. Then I have investigated problem with findings of webrtcendpoint in gsreamer registry.

## How to reproduce?

Build Kurento (`kms-omni-build` c7fbf099ece33995db5d06d22d8da967c0c532be Prepare release 6.16.0-0kurento1) against mainline supported gstreamer libraries 1.0 1.18.3

```
cmake -DCMAKE_INSTALL_PREFIX:PATH=/opt/kurento ../../git/kms-omni-build
make
make install
```

Run Kurento from local installation

```
#!/bin/sh

KURENTO_PREFIX="$(dirname "$(dirname "$(realpath "$0")")")"

if [ -e $KURENTO_PREFIX/lib/gstreamer-1.0 ] ; then
  if [ -n "$GST_PLUGIN_PATH" ] ; then
    GST_PLUGIN_PATH=":$GST_PLUGIN_PATH"
  fi

  export GST_PLUGIN_PATH="$KURENTO_PREFIX/lib/gstreamer-1.0$GST_PLUGIN_PATH"
fi

if [ -n "$LD_LIBRARY_PATH" ] ; then
  LD_LIBRARY_PATH=":$LD_LIBRARY_PATH"
fi

export LD_LIBRARY_PATH="$KURENTO_PREFIX/lib$LD_LIBRARY_PATH"

ldd -r $KURENTO_PREFIX/bin/kurento-media-server

export GST_DEBUG='*:4'

#gdb -args $KURENTO_PREFIX/bin/kurento-media-server -f "$KURENTO_PREFIX/etc/kurento/kurento.conf.json" -c "$KURENTO_PREFIX/etc/kurento/modules/kurento"  "$@
"
#strace -ff -o kurento-strace $KURENTO_PREFIX/bin/kurento-media-server -f "$KURENTO_PREFIX/etc/kurento/kurento.conf.json" -c "$KURENTO_PREFIX/etc/kurento/mo
dules/kurento"  "$@"
$KURENTO_PREFIX/bin/kurento-media-server -f "$KURENTO_PREFIX/etc/kurento/kurento.conf.json" -c "$KURENTO_PREFIX/etc/kurento/modules/kurento"  "$@"
```

To check gstreamer state

```
rm ~/.cache/gstreamer-1.0/registry.x86_64.bin
LD_LIBRARY_PATH=/opt/kurento/lib GST_PLUGIN_PATH=/opt/kurento/lib/gstreamer-1.0 gst-inspect-1.0
```

## Expected & current behavior

Run Kurento on regular system with distribution provided libraries in the future.

By actual hope is to replace need to run local ffplay on our lab RTMP stream with WebRTC technology.

Long term goal, maintainnambel BigBluButton.

I would report this issue and resolution and partil progress to BigBluButton inssue tracker for reference and possible cooperation.

Decide, please, if the library name or plugin entry point functions should be changes. Probably library name is better if you do not intend to push webrtc lugin to gstreamer mainline. Vice versa if you intend.

## (Optional) Possible solution

Desxribed above

## Info about your environment

Debian buster with backports repository. Kernel

```inux hathi 5.10.0-0.bpo.3-amd64 #1 SMP Debian 5.10.13-1~bpo10+1 (2021-02-11) x86_64 GNU/Linux```

### About Kurento Media Server

* Kurento version:  6.16.0-0kurento1 c7fbf099ece33995db5d06d22d8da967c0c532be
* Server OS:          Debian 10.9 with backports and gstreamer 1.0 1.18 from bullseye
* Installation method:
  - [] [apt-get]
  - [] [Docker]
  - [] [AWS CloudFormation]
  - [x] [Built from sources]

[apt-get]: https://doc-kurento.readthedocs.io/en/latest/user/installation.html#installation-local
[Docker]: https://doc-kurento.readthedocs.io/en/latest/user/installation.html#installation-docker
[AWS CloudFormation]: https://doc-kurento.readthedocs.io/en/latest/user/installation.html#installation-aws
[Built from sources]: https://doc-kurento.readthedocs.io/en/latest/dev/dev_guide.html#dev-sources

### About your Application Server

* Programming Language:  Java
* Kurento Client version:     kurento-hello-world 6.16.0 adf43cb3e5777316c5a7341a8347b9177647fc73


### About end-user clients

* Device(s):    Laptop,  ID 046d:0843 Logitech, Inc. Webcam C930e
* OS(es):        Debian 10.9 x86_64
* Browser(s):  Firefox 78.8.0esr (64-bit)

### Run these commands

```
cat: /etc/lsb-release: No such file or directory
```

```
/opt/kurento/bin/kurento-media-server -f /opt/kurento/etc/kurento/kurento.conf.json -c /opt/kurento/etc/kurento/modules/kurento  --version
Kurento Media Server version: 6.16.0
Found modules:
        'core' version 6.16.0
        'elements' version 6.16.0
        'filters' version 6.16.0
```

```
dpkg -l | grep -Pi 'kurento|kms-|gst.*1.5|nice'
ii  gir1.2-nice-0.1:amd64                         0.1.16-1~bpo10+1                             amd64        ICE library (GObject introspection)
ii  kms-cmake-utils                               6.16.1                                       all          Kurento CMake utils
ii  kms-jsonrpc                                   6.16.1                                       amd64        Kurento JSON-RPC library
ii  kms-jsonrpc-dev                               6.16.1                                       amd64        Kurento JSON-RPC library
ii  kmsjsoncpp                                    1.6.3                                        amd64        Kurento jsoncpp library
ii  kmsjsoncpp-dev                                1.6.3                                        amd64        Kurento jsonrpc library
rc  libgstreamer0.10-0:amd64                      0.10.36-1.5                                  amd64        Core GStreamer libraries and elements
rc  libgstreamer1.5-dev                           1.8.1                                        amd64        GStreamer core development files
ii  libnice-dev                                   0.1.16-1~bpo10+1                             amd64        ICE library (development files)
ii  libnice10:amd64                               0.1.16-1~bpo10+1                             amd64        ICE library (shared library)
ii  libreoffice-avmedia-backend-gstreamer         1:6.1.5-3+deb10u7                            amd64        GStreamer backend for LibreOffice
```

```
dpkg -l | grep -Pi 'gst.*1.0'
ii  gir1.2-gst-plugins-bad-1.0:amd64              1.18.3-1                                     amd64        GObject introspection data for the GStreamer libraries from the "bad" set
ii  gir1.2-gst-plugins-base-1.0:amd64             1.18.3-1                                     amd64        GObject introspection data for the GStreamer Plugins Base library
ii  gir1.2-gstreamer-1.0:amd64                    1.18.3-1                                     amd64        GObject introspection data for the GStreamer library
ii  gstreamer1.0-gl:amd64                         1.18.3-1                                     amd64        GStreamer plugins for GL
ii  gstreamer1.0-libav:amd64                      1.18.3-1                                     amd64        ffmpeg plugin for GStreamer
ii  gstreamer1.0-plugins-base:amd64               1.18.3-1                                     amd64        GStreamer plugins from the "base" set
ii  gstreamer1.0-plugins-good:amd64               1.18.3-1                                     amd64        GStreamer plugins from the "good" set
ii  gstreamer1.0-plugins-ugly:amd64               1.14.4-1                                     amd64        GStreamer plugins from the "ugly" set
ii  gstreamer1.0-pulseaudio:amd64                 1.18.3-1                                     amd64        GStreamer plugin for PulseAudio (transitional package)
ii  gstreamer1.0-tools                            1.18.3-1                                     amd64        Tools for use with GStreamer
ii  gstreamer1.0-tools-dbgsym                     1.18.3-1                                     amd64        debug symbols for gstreamer1.0-tools
ii  gstreamer1.0-x:amd64                          1.18.3-1                                     amd64        GStreamer plugins for X11 and Pango
ii  libgstreamer-gl1.0-0:amd64                    1.18.3-1                                     amd64        GStreamer GL libraries
ii  libgstreamer-opencv1.0-0:amd64                1.18.3-1                                     amd64        GStreamer OpenCV libraries
ii  libgstreamer-plugins-bad1.0-0:amd64           1.18.3-1                                     amd64        GStreamer libraries from the "bad" set
ii  libgstreamer-plugins-bad1.0-dev:amd64         1.18.3-1                                     amd64        GStreamer development files for libraries from the "bad" set
ii  libgstreamer-plugins-base1.0-0:amd64          1.18.3-1                                     amd64        GStreamer libraries from the "base" set
ii  libgstreamer-plugins-base1.0-dev:amd64        1.18.3-1                                     amd64        GStreamer development files for libraries from the "base" set
ii  libgstreamer1.0-0:amd64                       1.18.3-1                                     amd64        Core GStreamer libraries and elements
ii  libgstreamer1.0-0-dbgsym:amd64                1.18.3-1                                     amd64        debug symbols for libgstreamer1.0-0
ii  libgstreamer1.0-dev:amd64                     1.18.3-1                                     amd64        GStreamer core development files
ii  libgstreamer1.0-dev-dbgsym:amd64              1.18.3-1                                     amd64        debug symbols for libgstreamer1.0-dev
```

**Comment by micaelgallego** [comment]
The preferred way to deploy Kurento is in Docker containers. 

In any case, Kurento Media Server is supported in ubuntu 18.04.

Also, you can take a look to this branch https://github.com/Kurento/kurento-media-server/tree/bionic-gstreamer

**Comment by ppisa** [comment]
Thanks much for the reply. I understand that Docker is a way for stable system in production but I like native setup and packaging for development.

I have noticed bionic-gstreamer branch earlier but because it looked as updated last more than year ago, I did not investigated it too much till mow. As I see it is gstreamer1.0 1.14.0 based which would be ideal for stable Debian but on the other hand important webrtc extensions prepared as addons for gstreamer1.5 1.8 has not been included in 1.14 version yet and require supplements from other packages. 1.18 seems to provide what is required out of the box. Yes, probably there can be expected significant runtime mismatches....

The bionic-gstreamer branch requires version 7.0.0 of kms-cmake-utils which could be incompatible with my actual 6.x.x attempt. So I need separate build environment.....

Are there some plans for 7.x development continuation? It seems that it s only time from time updated from master and last update is quite old. There seems to be patch for names correction

```
First version of GStreamer 1.14 port

Updated GStreamer libs to 1.14
Renamed library/plugin filenames to match plugin's name (see http://gstreamer-devel.966125.n4.nabble.com/Plugin-loading-fails-with-Gstreamer-1-14-0-td4686497.html
Fixed possible SEGFAULT In kmsrtpendpoint
```

I need to check if it works.

Thanks again for reply,

Pavel

**Comment by dangersvn** [comment]
Hi @ppisa,
Did it work for you?
I'm interested in how you upgrade GStreamer version in the Kurento server and build the Kurento docker image with a specific version of GStreamer. 
Looking forward to hearing from your insight.

Thanks,
Dang

**Comment by j1elo** [comment]
We're just finishing the release of Kurento 6.18 (only compatible with Ubuntu 16.04, Ubuntu 18.04, and our old fork of GStreamer 1.8)

Next thing is focusing on Kurento 7.0, which at last, drops all dependencies on our old GStreamer forks, and is able to compile and run with the native GStreamer packages that come by default with Ubuntu 20.04 (GStreamer 1.16). This is a big step because now Kurento is able to work with a much more modern version and we can stop maintaining our forks...

I tried building the 7.0 branch on a newer Ubuntu 22.04 (that comes with GStreamer 1.20) but it breaks everywhere. I think there were some internal changes between GStreamer 1.16 and 1.20, that makes Kurento break apart. So, one step is releasing Kurento 7.0 for Ubuntu 20.04, and then a second step for the distant future is to see what happened in GStreamer 1.20 and what needs to be done to make Kurento work with it.

**Comment by j1elo** [comment]
This was fixed for 7.0.0, potentially related commits:
* https://github.com/Kurento/kms-core/commit/08b1b36ec9e592da49277ea67d8988038640fee5
* https://github.com/Kurento/kms-elements/commit/342f4e4d1c0cd640928ad8f725b78a0f333a1b6e
